### ISSUE TITLE
The implementation of perceiver doesn't seem to correspond to what is described in the paper

### ISSUE THREAD

**Issue by hr0nix** [deepmind/deepmind-research#253]
Hi!

There seems to be just one cross-attention op before all self-attention ops, not cross-attention in every block. Is this a bug, by design or am I reading the code incorrectly?

Relevant line:
https://github.com/deepmind/deepmind-research/blob/2fede222dd33545f0a7b91a9afd67ca957009ba6/perceiver/perceiver.py#L467

**Comment by seb5666** [comment]
Hi!

This is indeed by design. If you look at Figure 2 of the paper, you can see that the cross-attention operation is only part of the "Encode"  block. The "Process" block (which is repeated L times) only has a self-attention operation.

**Comment by djaegle** [comment]
Hi, thanks for your comment! The figure Seb is referring to is in the Perceiver IO paper (https://arxiv.org/abs/2107.14795) - we'll be updating the README shortly with this link. We found that we can get better performance on ImageNet than reported in the original Perceiver paper with only a single cross-attention, so this is the model we've included here. The models described in the Perceiver IO paper perform better than their counterparts in the original paper and are roughly half as expensive FLOPs-wise because they use only one cross-attend.

Sorry for any confusion!
